trio/_dtls.py

```diff
@@ -0,0 +1,959 @@
+
+    # HelloVerifyRequest body is:
+    # - 2 bytes version
+    # - length-prefixed cookie
+    #
+    # The DTLS 1.2 spec says that for this message specifically we should use
+    # the DTLS 1.0 version.
+    #
+    # (It also says the opposite of that, but that part is a mistake:
+    #    https://www.rfc-editor.org/errata/eid4103
+    # ).
+    #
+    # And I guess we use this for both the message-level and record-level
+    # ProtocolVersions, since we haven't negotiated anything else yet?
+    body = ProtocolVersion.DTLS10 + bytes([len(cookie)]) + cookie
+
+    # RFC says have to copy the client's record number
+    # Errata says it should be handshake message number
+    # Openssl copies back record sequence number, and always sets message seq
+    # number 0. So I guess we'll follow openssl.
+    hs = HandshakeFragment(
+        msg_type=HandshakeType.hello_verify_request,
+        msg_len=len(body),
+        msg_seq=0,
+        frag_offset=0,
+        frag_len=len(body),
+        frag=body,
+    )
+    payload = encode_handshake_fragment(hs)
+
+    packet = encode_record(
+        Record(ContentType.handshake, ProtocolVersion.DTLS10, epoch_seqno, payload)
+    )
+    return packet
+
+
+class _Queue:
+    def __init__(self, incoming_packets_buffer):
+        self.s, self.r = trio.open_memory_channel(incoming_packets_buffer)
+
+
+def _read_loop(read_fn):
+    chunks = []
+    while True:
+        try:
+            chunk = read_fn(2 ** 14)  # max TLS record size
+        except SSL.WantReadError:
+            break
+        if not chunk:
+            break
+        chunks.append(chunk)
+    return b"".join(chunks)
+
+
+async def handle_client_hello_untrusted(dtls, address, packet):
+    if dtls._listening_context is None:
+        return
+
+    try:
+        epoch_seqno, cookie, bits = decode_client_hello_untrusted(packet)
+    except BadPacket:
+        return
+
+    if not valid_cookie(cookie, address, bits):
+        challenge_packet = challenge_for(address, epoch_seqno, bits)
+        try:
+            async with dtls._send_lock:
+                await dtls.socket.sendto(challenge_packet, address)
+        except (OSError, trio.ClosedResourceError):
+            pass
+    else:
+        # We got a real, valid ClientHello!
+        stream = DTLSStream._create(dtls, address, dtls._listening_context)
+        try:
+            stream._inject_client_hello_untrusted(packet)
+        except BadPacket:
+            # ...or, well, OpenSSL didn't like it, so I guess we didn't.
+            return
+        old_stream = dtls._streams.get(address)
+        if old_stream is not None:
+            if old_stream._client_hello == packet:
+                # ...but it's just a duplicate of a packet we got before, so never mind.
+                return
+            else:
+                # Ok, this *really is* a new handshake; the old stream should go away.
+                old_stream._replaced()
```

Comment:
Got an error in this code path about bool not being callable, line 709 potentially a method/attribute name collision